### PR TITLE
fix(train): fix regression from #971 - keep Muon-excluded matrices at Muon's LR

### PR DESCRIPTION
**Pull request by WindChimeRan**
## Purpose

Fixes a regression introduced by #971.

* See the DSpark x Markov head report: https://claude.ai/code/artifact/124e424c-9db1-42bd-a724-7d5657f59df7
* See the DFlash2 codebook report: https://claude.ai/code/artifact/380c43b0-5609-4676-bc2f-68c338b2def5

#971 correctly routed DSpark's Markov factors (`markov_w1`, `markov_w2`) out of Muon — they are vocabulary-indexed tables, and orthogonalizing one is meaningless. But `build_optimizers` has a single AdamW group, at `lr`. So the factors silently dropped from `muon_lr` / `muon_weight_decay` to `lr` / `weight_decay` — **10x lower on both**, given `muon_lr = 10 * lr` by default. Nothing in the diff said so.

The head then barely trains. Over 10 epochs its logit bias moves from 0.006 to 0.027, against 0.587 before #971 — it isn't learning something worse, it isn't learning.

## Change

The 2D matrices excluded from Muon for semantic reasons share a single AdamW optimizer with the base parameters, using a dedicated parameter group at `muon_lr` / `muon_weight_decay`.

This covers `markov_w1` / `markov_w2` and the DFlash2 codebooks. `embed_tokens` / `lm_head` are frozen in every speculator type, so no other model changes. Everything else from #971 is untouched — the `std=0.01` init stays, and both factors stay excluded from Muon.

## Results

Qwen3-8B, `tutorial_regen` (4993 rows), 6xH100, 10 epochs, single seed. Peak validation acceptance length:

|  | peak val EAL | Δ |
|---|---|---|
| pre-#971 | 2.552 | — |
| `main` (#971) | 2.497 | **−0.055** |
| this PR | 2.552 | recovered |

`main` is below pre-#971 at all ten epochs, by 0.049 to 0.072.

## Test

`test_vocab_factors_keep_muon_lr_not_the_base_lr` fails on `main` with `assert 0.0003 == 0.003` and passes here. Partitioning tests also cover the DFlash2 transition codebooks and verify that generic embeddings, LM heads, and unrelated codebooks retain base AdamW hyperparameters. Full unit suite shows no new failures.

## Note on resuming older checkpoints

The AdamW parameter-group layout changes, so checkpoints written before this PR are not guaranteed to resume. This is deliberate: those checkpoints come from runs whose Markov head was inert, and any incompatibility fails loudly rather than silently mis-mapping optimizer state.

Fresh runs and `--optimizer adamw` are unaffected.

## Checklist

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan/results, such as providing test command and pasting the results.
- [ ] (Optional) The necessary documentation update.
- [x] I (a human) have written or reviewed the code in this pr to the best of my ability.

### CODE REVIEW

File: src/speculators/train/optimizers.py [hunk 1]
```diff
@@ -111,6 +120,15 @@ def build_optimizers(model: Module, config) -> list[torch.optim.Optimizer]:
...
             )
+        if excluded:
+            # Skipped by Muon for their shape, not to be trained more slowly.
+            optimizers.append(
+                torch.optim.AdamW(
+                    excluded,
+                    lr=config.muon_lr,
+                    weight_decay=config.muon_weight_decay,
+                )
+            )
```
[the remark this answers is not in the document]
Comment (WindChimeRan) [reply src/speculators/train/optimizers.py]: This is a pre-existing over-defensive non-issue.

**Comment by fynnsu** [comment]
I don't love that this introduces a third optimizer group. Also my understanding is that the lr for muon can typically be higher because optimization should be more stable.

Obviously it's not good that the `markov_w1` and `markov_w2` weights aren't being trained in this example, but perhaps the learning rate in general needs to be higher (instead of specifically the lr for these weights)?

What do you think?

**Comment by WindChimeRan** [comment]
@fynnsu 
> I don't love that this introduces a third optimizer group.

Do you mean the code style or the algorithm correctness? 

> lr for muon can typically be higher because optimization should be more stable.

I agree, and it should be aligned with the batch size, you can tune `config.muon_lr` and `config.lr` according to batch size. 

> the learning rate in general needs to be higher (instead of specifically the lr for these weights)

I didn't specify lr for these weights, it's `config.muon_lr`

**Comment by WindChimeRan** [comment]
updates: 
* refactoring: make the code slightly cleaner and future proof
* dflash2: extend the fix to dflash as well: ablation https://claude.ai/code/artifact/380c43b0-5609-4676-bc2f-68c338b2def5?open_in_browser=1&via=user_open&org=4d8a5f93-b754-484f-972e-39bcb25b9848

**Comment by fynnsu** [comment]
Hi @WindChimeRan, I've been working on a system for running training/ablations cleanly on the clusters and decided to test it on this pr. I ran the following test:

Qwen3-8B, on the same 5k tutorial_regen dataset, 3 epochs. Three runs:
1. (YELLOW) latest commit from main branch. Default lr settings (`lr=1e-4`, `muon_lr=1e-3`)
2. (GREEN) Merge commit of this pr + latest commit on main. Default lr settings (`lr=1e-4`, `muon_lr=1e-3`, `markov_lr=1e-3`)
3. (RED) latest commit from main branch. Both `lr=1e-3` and `muon_lr=1e-3`

Note: that the first two are basically the same as your "main" and "this pr" runs. Whereas the 3 just asks "what if we set both of the learning rates to the same higher value". 




image


You can see above, that 3 (RED) matches the performance of 2 (GREEN) as the model trains for 3 epochs. It seems do pretty much the same, maybe marginally better on a few graphs.

You can also very clearly see the reproduction of the issue with the main branch alone 1 (YELLOW), which exactly matches the performance of the other runs at `position_0_acc` but lags at later positions, which confirms our hypothesis that the markov correction head is not learning. 

Based on this, I don't believe this pr (and a separate optimizer group for markov heads alone) is strictly necessary. Instead, I wonder if we should be raising our default `lr` and/or having it match the `muon_lr` instead of being 10x smaller. What do you think?

**Comment by WindChimeRan** [comment]
@fynnsu 

Thanks for the ablation. I agree we can go with run 3. 

---

Just need to fix all the default settings in examples, in CI, in doc tutorial. and maybe add a comment in code to explain this.
